modules/miner/blockmanager.go

```diff
@@ -1,127 +1,115 @@
 		delete(m.arbDataMem, m.headerMem[m.memProgress])
 		m.memProgress++
-		if m.memProgress == headerForWorkMemory {
+		if m.memProgress == HeaderMemory {
 			m.memProgress = 0
 		}
 	}
+
+	// Update the source block.
+	block := m.blockForWork()
+	m.sourceBlock = &block
+	m.sourceBlockTime = time.Now()
 }
 
-// HeaderForWork returns a block that is ready for nonce grinding, along with
-// the root hash of the block.
+// HeaderForWork returns a header that is ready for nonce grinding. The miner
+// will store the header in memory for a while, depending on the constants
+// 'HeaderMemory', 'BlockMemory', and 'MaxSourceBlockAge'. On the full network,
+// it is typically safe to assume that headers will be remembered for
+// min(10 minutes, 1000 requests).
 func (m *Miner) HeaderForWork() (types.BlockHeader, types.Target, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Return a blank header with an error if the wallet is locked.
 	if !m.wallet.Unlocked() {
 		return types.BlockHeader{}, types.Target{}, modules.ErrLockedWallet
 	}
-	m.mu.Lock()
-	defer m.mu.Unlock()
+
+	// Check that the wallet has been initialized, and that the miner has
+	// successfully fetched an address.
 	err := m.checkAddress()
 	if err != nil {
 		return types.BlockHeader{}, types.Target{}, err
 	}
 
-	if time.Since(m.lastBlock).Seconds() > secondsBetweenBlocks {
-		m.prepareNewBlock()
+	// If too much time has elapsed since the last source block, get a new one.
+	// This typically only happens if the miner has just turned on after being
+	// off for a while. If the current block has been used for too many
+	// requests, fetch a new source block.
+	if time.Since(m.sourceBlockTime) > MaxSourceBlockAge || m.memProgress%(HeaderMemory/BlockMemory) == 0 {
+		m.newSourceBlock()
 	}
 
-	// The header that will be returned for nonce grinding.
-	// The header is constructed from a block and some arbitrary data. The
-	// arbitrary data allows for multiple unique blocks to be generated from
-	// a single block in memory. A block pointer is used in order to avoid
-	// storing multiple copies of the same block in memory
-	var header types.BlockHeader
-	var arbData []byte
-	var block *types.Block
-
-	if m.memProgress%(headerForWorkMemory/blockForWorkMemory) == 0 {
-		// Grab a new block. Allocate space for the pointer to store it as well
-		block = new(types.Block)
-		*block, _ = m.blockForWork()
-		header = block.Header()
-		arbData = block.Transactions[0].ArbitraryData[0]
-
-		m.lastBlock = time.Now()
-	} else {
-		// Set block to previous block, but create new arbData
-		block = m.blockMem[m.headerMem[m.memProgress-1]]
-		arbData, _ = crypto.RandBytes(types.SpecifierLen)
-		block.Transactions[0].ArbitraryData[0] = arbData
-		header = block.Header()
+	// Create a header from the source block - this may be a race condition,
+	// but I don't think so (underlying slice may be shared with other blocks
+	// accessible outside the miner).
+	var arbData [crypto.EntropySize]byte
```

Comment:
pointers =/

The arb data gets manipulated frequently. If you use a slice, 'pass-by-value' means the underlying pointer is still pointed at the same data, which was causing incorrect block reassembly.